url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/internal/chartpart/AxisPair.java

```diff
@@ -0,0 +1,228 @@
+import org.knowm.xchart.internal.Series;
+import org.knowm.xchart.internal.Series_AxesChart;
+import org.knowm.xchart.style.AxesChartStyler;
+import org.knowm.xchart.style.CategoryStyler;
+
+/**
+ * @author timmolter
+ */
+public class AxisPair<ST extends AxesChartStyler, S extends Series> implements ChartPart {
+
+  private final Chart<AxesChartStyler, Series_AxesChart> chart;
+
+  private final Axis<AxesChartStyler, Series_AxesChart> xAxis;
+  private final Axis<AxesChartStyler, Series_AxesChart> yAxis;
+
+  /**
+   * Constructor
+   *
+   * @param chart
+   */
+  public AxisPair(Chart<AxesChartStyler, Series_AxesChart> chart) {
+
+    this.chart = chart;
+
+    // add axes
+    xAxis = new Axis<AxesChartStyler, Series_AxesChart>(chart, Axis.Direction.X);
+    yAxis = new Axis<AxesChartStyler, Series_AxesChart>(chart, Axis.Direction.Y);
+  }
+
+  @Override
+  public void paint(Graphics2D g) {
+
+    prepareForPaint();
+
+    yAxis.paint(g);
+    xAxis.paint(g);
+  }
+
+  private void prepareForPaint() {
+
+    // set the axis data types, making sure all are compatible
+    xAxis.setAxisDataType(null);
+    yAxis.setAxisDataType(null);
+    for (Series_AxesChart series : chart.getSeriesMap().values()) {
+      xAxis.setAxisDataType(series.getxAxisDataType());
+      yAxis.setAxisDataType(series.getyAxisDataType());
+    }
+
+    // calculate axis min and max
+    xAxis.resetMinMax();
+    yAxis.resetMinMax();
+
+    // if no series, we still want to plot an empty plot with axes. Since there are no min and max with no series added, we just fake it arbitrarily.
+    if (chart.getSeriesMap() == null || chart.getSeriesMap().size() < 1) {
+      xAxis.addMinMax(-1, 1);
+      yAxis.addMinMax(-1, 1);
+    }
+    else {
+      for (Series_AxesChart series : chart.getSeriesMap().values()) {
+        // add min/max to axes
+        // System.out.println(series.getxMin());
+        // System.out.println(series.getxMax());
+        // System.out.println(series.getyMin());
+        // System.out.println(series.getyMax());
+        // System.out.println("****");
+        xAxis.addMinMax(series.getXMin(), series.getXMax());
+        yAxis.addMinMax(series.getYMin(), series.getYMax());
+      }
+    }
+
+    overrideMinMax();
+
+    // logarithmic sanity check
+    if (chart.getStyler().isXAxisLogarithmic() && xAxis.getMin() <= 0.0) {
+      throw new IllegalArgumentException("Series data (accounting for error bars too) cannot be less or equal to zero for a logarithmic X-Axis!!!");
+    }
+    if (chart.getStyler().isYAxisLogarithmic() && yAxis.getMin() <= 0.0) {
+      // System.out.println(getMin());
+      throw new IllegalArgumentException("Series data (accounting for error bars too) cannot be less or equal to zero for a logarithmic Y-Axis!!!");
+    }
+  }
+
+  /**
+   * Here we can add special case min max calculations and take care of manual min max settings.
+   */
+  public void overrideMinMax() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 183-207 into a separate method.